zerver/actions/message_edit.py

```diff
@@ -1296,6 +1268,83 @@ def check_time_limit_for_change_all_propagate_mode(
     )
 
 
+def get_message_edit_request_object(
+    message: Message,
+    user_profile: UserProfile,
+    propagate_mode: str,
+    stream_id: int | None = None,
+    topic_name: str | None = None,
+    content: str | None = None,
+) -> StreamMessageEditRequest | DirectMessageEditRequest:
+    if not message.is_stream_message():
+        # We have already validated the code to have content
+        # as not None.
+        assert content is not None
+        return DirectMessageEditRequest(
+            content=content,
+            orig_content=message.content,
+            is_content_edited=True,
+        )
+
+    is_content_edited = False
+    new_content = message.content
+    if content is not None:
+        is_content_edited = True
+        if content.rstrip() == "":
+            content = "(deleted)"
+        new_content = normalize_body(content)
+
+    is_topic_edited = False
+    topic_resolved = False
+    topic_unresolved = False
+    old_topic_name = message.topic_name()
+    new_topic_name = old_topic_name
+    pre_truncation_new_topic_name = old_topic_name
+
+    if topic_name is not None:
+        is_topic_edited = True
+        pre_truncation_new_topic_name = topic_name
+        new_topic_name = truncate_topic(topic_name)
+
+        resolved_prefix_len = len(RESOLVED_TOPIC_PREFIX)
+        topic_resolved = (
+            new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+            and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+            and pre_truncation_new_topic_name[resolved_prefix_len:] == old_topic_name
+        )
+        topic_unresolved = (
+            old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+            and not new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+            and old_topic_name.lstrip(RESOLVED_TOPIC_PREFIX) == new_topic_name
+        )
+
+    orig_stream_id = message.recipient.type_id
+    orig_stream = get_stream_by_id_in_realm(orig_stream_id, message.realm)
+
+    is_stream_edited = False
+    new_stream = orig_stream
+    if stream_id is not None:
+        new_stream = access_stream_by_id(user_profile, stream_id, require_active=True)[0]
```

Comment:
Fetching the stream here means that validation of the new stream ID is done here and invalid or inaccessible stream ID error would be raised first before other errors, which was not the case before. But this seems better than just setting the ID here and then have the `stream` field optional and then using assert statements to make sure that the `stream` field is not None.